networkx/algorithms/flow/dinitz_alg.py

```diff
@@ -173,35 +173,64 @@ def dinitz_impl(G, s, t, capacity, residual, cutoff):
 
     def breath_first_search():
         parents = {}
-        queue = deque([s])
+        vertex_dist = {s: 0}
+        queue = deque([(s, 0)])
+        # Record all the potential edges of shortest augmenting paths
         while queue:
             if t in parents:
                 break
-            u = queue.popleft()
-            for v in R_succ[u]:
-                attr = R_succ[u][v]
-                if v not in parents and attr["capacity"] - attr["flow"] > 0:
-                    parents[v] = u
-                    queue.append(v)
+            u, dist = queue.popleft()
+            for v, attr in R_succ[u].items():
+                if attr["capacity"] - attr["flow"] > 0:
+                    if v in parents:
+                        if vertex_dist[v] == dist + 1:
+                            parents[v].append(u)
+                    else:
+                        parents[v] = deque([u])
+                        vertex_dist[v] = dist + 1
+                        queue.append((v, dist + 1))
         return parents
 
     def depth_first_search(parents):
+        # DFS to find all the shortest augmenting paths
         """Build a path using DFS starting from the sink"""
-        path = []
+        total_flow = 0
         u = t
-        flow = INF
-        while u != s:
-            path.append(u)
-            v = parents[u]
-            flow = min(flow, R_pred[u][v]["capacity"] - R_pred[u][v]["flow"])
+        # path also functions as a stack
+        path = [u]
+        # The loop ends with no augmenting path left in the layered graph
+        while True:
+            if len(parents[u]) > 0:
+                v = parents[u][0]
+                path.append(v)
+            else:
+                path.pop()
+                if len(path) == 0:
+                    break
+                v = path[-1]
+                parents[v].popleft()
+            # Augment the flow along the path found
+            if v == s:
+                flow = INF
+                for u, v in pairwise(path):
+                    flow = min(flow, R_pred[u][v]["capacity"] - R_pred[u][v]["flow"])
+                u_index = len(path) - 2
+                while u_index >= 0:
+                    u = path[u_index]
+                    v_index = u_index + 1
+                    v = path[v_index]
```

Comment:
This seems quite cumbersome.  Along with line 229, can this be replaced by a call like:
```python
for u, v in pairwise(reversed(path)):
```
I do see that you are cutting off the top of the path within this loop, but presumably that won't impact the for loop implementation because it doesn't influence the index system.
Is there an advantage to using the indexes that I'm not seeing?